### PR TITLE
[Flax] Add docstrings & model outputs

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR adds docstring examples & `ModelOutputs` to Flax. This includes `all_hidden_states`.
The code necessary for `all_attentions` is added as well, but it requires a change in the official Flax library.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 1]
```diff
@@ -456,7 +570,28 @@ def __call__(
...
+
+        # make sure that model does not return dict when in jitted mode
+        if not isinstance(input_ids, (DeviceArray, np.ndarray)) and return_dict:
+            logger.warn(
+                "Model cannot return dictionary in jitted mode. Setting `return_dict=False`... To suppress this warning, please set ``return_dict=False``."
+            )
+            return_dict = False
+
+        if output_attentions:
+            raise NotImplementedError(
```
Comment (patrickvonplaten) [inline src/transformers/models/bert/modeling_flax_bert.py:591]: @sgugger @marcvanzee @avital - For now it is not possible to return attentions because Flax `flax.linen.dot_product_attention` does not allow it. Not sure what the best approach is here - could we maybe adapt `dot_product_attention` slightly to also return attention_weights with a flag?

File: tests/test_modeling_flax_common.py [hunk 2]
```diff
@@ -195,19 +237,47 @@ def test_jit_compilation(self):
...
+                @jax.jit
+                def model_jitted_return_dict(input_ids, attention_mask=None, token_type_ids=None):
+                    return model(
+                        input_ids=input_ids,
+                        attention_mask=attention_mask,
+                        token_type_ids=token_type_ids,
+                    )
+
+                # jitted function cannot return OrderedDict
+                with self.assertRaises(TypeError):
```
Comment (patrickvonplaten) [inline tests/test_modeling_flax_common.py:266]: Jitted function cannot return OrderedDict => this shouldn't really be a problem though

File: src/transformers/file_utils.py [hunk 3]
```diff
@@ -915,30 +915,129 @@ def _prepare_output_docstrings(output_type, config_class):
...
+
+        if model_class[:2] == "TF":
+            prefix = "TF_"
+        elif model_class[:4] == "Flax":
+            prefix = "FLAX_"
+        else:
+            prefix = "PT_"
+
         doc_kwargs = dict(model_class=model_class, tokenizer_class=tokenizer_class, checkpoint=checkpoint)
+        this_module = sys.modules[__name__]
```
Comment (sgugger) [inline src/transformers/file_utils.py:1024]: Not super fan of the use of `sys.modules` and getattr. An explicit mapping takes a little bit more lines of code but might be cleaner? It's what we do for configs or tokenizers.

File: src/transformers/modeling_flax_outputs.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+from .file_utils import ModelOutput
+
+
+@dataclass
+class FlaxBaseModelOutput(ModelOutput):
+    """
+    Base class for model's outputs, with potential hidden states and attentions.
+
+    Args:
+        last_hidden_state (:obj:`DeviceArray` of shape :obj:`(batch_size, sequence_length, hidden_size)`):
```
Comment (sgugger) [inline src/transformers/modeling_flax_outputs.py:29]: We usually say `torch.Tensor` or `tf.Tensor`, so it  would be nice to have a full path. At the same time, `jaxlib.xla_extension.DeviceArray` is a mouthful. Is there a common abbreviation of `jaxlib.xla_extension` we could use?

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 5]
```diff
@@ -456,7 +570,21 @@ def __call__(
...
     ):
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        return_dict = return_dict if return_dict is not None else self.config.return_dict
+
+        if output_attentions:
+            raise NotImplementedError(
+                "Currently attention scores cannot be returned." "Please set `output_attentions` to False for now."
```
Comment (sgugger) [inline src/transformers/models/bert/modeling_flax_bert.py:585]:
```suggestion
                "Currently attention scores cannot be returned. Please set `output_attentions` to False for now."
```

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 6]
```diff
@@ -493,17 +624,58 @@ def setup(self):
...
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+        )
+        hidden_states = outputs[0]
+        pooled = self.pooler(hidden_states) if self.add_pooling_layer else None
+
+        outputs = (
+            hidden_states,
+            pooled,
+        ) + outputs[1:]
```
Comment (sgugger) [inline src/transformers/models/bert/modeling_flax_bert.py:654]:
```suggestion
        outputs = (hidden_states, pooled) + outputs[1:]
```
This fits in one line ;-)

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 7]
```diff
@@ -324,8 +425,21 @@ class FlaxBertEncoder(nn.Module):
...
-        return self.layer(hidden_states, attention_mask, deterministic=deterministic)
+    def __call__(
+        self,
+        hidden_states,
+        attention_mask,
+        deterministic: bool = True,
+        output_attentions: bool = False,
+        output_hidden_states: bool = False,
+    ):
+        return self.layer(
```
Comment (sgugger) [inline src/transformers/models/bert/modeling_flax_bert.py:436]: Shouldn't we have the base output here with the `return_dict` argument?

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 8]
```diff
@@ -493,17 +624,58 @@ def setup(self):
...
+        else:
+            all_attentions = None
+
+        # extract all hidden_states
+        if output_hidden_states and output_attentions:
+            all_hidden_states = outputs[3]
+        elif output_hidden_states and not output_attentions:
+            all_hidden_states = outputs[2]
+        else:
+            all_hidden_states = None
```
Comment (sgugger) [inline src/transformers/models/bert/modeling_flax_bert.py:671]: This could be simplified a lot if the encoder was retuning a `ModelOutput`.

File: tests/test_modeling_flax_common.py [hunk 9]
```diff
@@ -85,6 +89,44 @@ def assert_almost_equals(self, a: np.ndarray, b: np.ndarray, tol: float):
...
+                        recursive_check(tuple_iterable_value, dict_iterable_value)
+                elif tuple_object is None:
+                    return
+                else:
+                    self.assertTrue(
+                        self.assert_almost_equals(
+                            set_nan_tensor_to_zero(tuple_object), set_nan_tensor_to_zero(dict_object), 1e-5
+                        ),
+                        msg=f"Tuple and dict output are not equal. Difference: {np.max(np.abs(tuple_object - dict_object))}. Tuple has `nan`: {np.isnan(tuple_object).any()} and `inf`: {np.isinf(tuple_object)}. Dict has `nan`: {np.isnan(dict_object).any()} and `inf`: {np.isinf(dict_object)}.",
+                    )
```
Comment (sgugger) [inline tests/test_modeling_flax_common.py:115]: `self.assert_almost_equals` should already do an assert no?

**Review by sgugger — APPROVED** [review]
Thanks a lot for working on this!

File: src/transformers/file_utils.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/transformers/file_utils.py:1024]: Ok for me! I do think though that the code is `file_utils.py` doesn't need to be as readable as the configuration or modeling code since fewer people will actually read the code that generates the docstrings.

File: src/transformers/models/bert/modeling_flax_bert.py (hunk 7 above)
Comment (patrickvonplaten) [reply src/transformers/models/bert/modeling_flax_bert.py:436]: Good point! That's what we do in `modeling_bert.py`, but I don't really agree with it - I think only `PreTrainedModel`' classes should have a `return_dict` input and all `flax.linen.Module` should always return tuples .... Happy to discuss though! What do you think?

Comment (patrickvonplaten) [reply src/transformers/models/bert/modeling_flax_bert.py:436]: Also cc @LysandreJik

Comment (patrickvonplaten) [reply src/transformers/models/bert/modeling_flax_bert.py:436]: Actually, I don't see big downside to not using `return_dict` => I'll just implement it

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 10]
```diff
@@ -239,13 +293,22 @@ def setup(self):
...
+        attn_output = attn_outputs[0]
         hidden_states = self.output(attn_output, hidden_states, deterministic=deterministic)
-        return hidden_states
+
+        outputs = (hidden_states,)
+
+        if output_attentions:
+            outputs += attn_outputs[1]
+
+        return outputs
```
Comment (patrickvonplaten) [inline src/transformers/models/bert/modeling_flax_bert.py:311]: Now we return tuples instead of tensors

File: src/transformers/models/bert/modeling_flax_bert.py (hunk 1 above)
Comment (avital) [reply src/transformers/models/bert/modeling_flax_bert.py:587]: Oh this was something I forgot to do! Let me to it today or tomorrow -- we can refactor `dot_product_attention` to be composed of two functions, one of which just returns the attention mask

**Review by LysandreJik — COMMENTED** [review]
Great!
